### ISSUE TITLE
How to improve performance on a single node

### ISSUE THREAD

**Issue by janvainer** [horovod/horovod#2380]
**Environment:**
1. Framework: PyTorch
2. Framework version: 1.7.0 dev
3. Horovod version: 0.20.0
4. MPI version: 3.1.4
5. CUDA version: 10.2
6. NCCL version: 2.3.5
7. Python version: 3.6.9
8. OS and version: Ubuntu bionic 18.04
9. GCC version: 7.5.0

**Checklist:**
I found similar issues but with regard to multi-node training.

**Your question:**
Hi, I am measuring horovod efficiency on a custom model on a server with 4 Nvidia T4 16GB GPUs and I am not able to get sastisfying results. I calculated how long it takes to run 100 gradient updates with 1,2,3 and 4 GPUs with batch size 30 on each of them. Then I calculated how many items the model sees per second. Finally, I compared how big is the training speedup in terms of seen data items. Below is a table with results and an efficiency plot. I am running my experiments with the following command. `horovodrun -np n_gpus python my_script.py --some_args`. What could be the reason for such an efficiency drop? With a single gpu, I am getting almost 100% GPU utilization and around 80% GPU memory usage. With more GPUs, the utilization drops to around 60% for any number of GPUs, while memory stays at 80%. 

I use the default pytorch DataLoader without memory pinning and with num_workers set to zero. What puzzles me is that the utilization does not drop when using 2 or more GPUs. If it did, I would suspect that IO is the problem.. I am running more experiments with memory pinning and more workers for the dataloader, but I was wondering is this sort of efficiency is normal or if there could be some problem with my horovod setup. Could you point me in some horovod specific directions how to improve the efficiency?

![obrazek](https://user-images.githubusercontent.com/31015638/96264455-7b34da00-0fc4-11eb-96f3-d4b8af590fa4.png)
![obrazek](https://user-images.githubusercontent.com/31015638/96264050-11b4cb80-0fc4-11eb-8d4e-4a16fb6e9a2d.png)

**Comment by tgaddair** [comment]
Hey @LordOfLuck, when running into these sorts of issues, the first thing we need to do is figure out is the problem is coming from the way Horovod is configured in your environment, or the way it's being used in the script.

To do that, can you try running our standard [pytorch_synthetic_benchmark.py](https://github.com/horovod/horovod/blob/master/examples/pytorch/pytorch_synthetic_benchmark.py) script with the same experiment setup and share the results?

**Comment by LordOfLuck** [comment]
Hi, thanks for response. I am sending the benchmark results for single gpu and 2-gpu training

Single GPU with horovod:
```
[1,0]<stdout>:Model: resnet50
[1,0]<stdout>:Batch size: 32
[1,0]<stdout>:Number of GPUs: 1
...
[1,0]<stdout>:Img/sec per GPU: 114.5 +-1.1
[1,0]<stdout>:Total img/sec on 1 GPU(s): 114.5 +-1.1
```
I also tried to remove horovod-related code from the benchmark script altogether and got basically the same performance on a single GPU.

Two GPUs
```
[1,0]<stdout>:Model: resnet50
[1,0]<stdout>:Batch size: 32
[1,0]<stdout>:Number of GPUs: 2
...
[1,0]<stdout>:Img/sec per GPU: 95.8 +-5.8
[1,0]<stdout>:Total img/sec on 2 GPU(s): 191.6 +-11.5
```
I also tried the same setup on 4 GPUs and got basically the same statistics

**Comment by tgaddair** [comment]
Hey @LordOfLuck, good catch regarding NCCL, that is definitely the likely issue.

I would suggest reinstalling Horovod with the environment variable `HOROVOD_GPU_OPERATIONS=NCCL` set.  That will ensure that NCCL support is built, or an error will be raised during install.

**Comment by LordOfLuck** [comment]
Thanks, I have one more issue. I sem to be getting much higher GPU utilization during validation (no gradient update) than during training. During training, I am getting around 60% utilization while during validation it is almost 100%. Is this drop normal? Could it be fully caused by the communication between processes, or is there something else? Any suggestions would be welcome, I will try them and report back.

**Comment by tgaddair** [comment]
Have you reinstalled Horovod with NCCL?  If not, that would explain the drop in util, as communication would become the bottleneck.  With NCCL, you should see close to 100% util during training.

Horovod + AMP should work fine.  We have a number of users who are using this combination.

**Comment by LordOfLuck** [comment]
Yes I did and It improved a bit, but the utilization remains low. Definitely far from 100%. The utilization looks as follows now:
![obrazek](https://user-images.githubusercontent.com/31015638/96896727-25fa3c00-148e-11eb-97d9-99906a65cbc6.png)

The spikes around 100% happen during evaluation, while the valleys correspond to training phase. It is better than without NCCL, but there still seems to be room for improvement. IO is probably out of question since the same data loading process is used in both phases. My simplified horovod related code is below:

```
torch.cuda.set_device(hvd.local_rank())
model.cuda()
model = convert_sync_batchnorm(model)

hvd.broadcast_parameters(self.model.state_dict(), root_rank=0)
optimizer = hvd.DistributedOptimizer(
      optimizer, named_parameters=model.named_parameters()
)
train_loader = make_distributed_loader(
      train_loader, hvd.size(), rank=hvd.local_rank()
 )
```

The rest is standard pytorch code. I am using gradient clipping, which could slow down all processes as it happens on each one separately. Could this be the reason?
